firebase_admin/app_check.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# See as an example from firebase_admin/messaging.py
+# def _get_messaging_service(app):
+#     return _utils.get_app_service(app, _MESSAGING_ATTRIBUTE, _MessagingService)
+
+# Our goal in general is to take the design doc implentation and match it to the 
+# existing code in the SDK with tests
+# Timeline is to be done by week of the 19th 
+
+"""Firebase App Check module."""
+
+# ASK(lahiru) Do I need to add these imports to the requirements file?
+import jwt
+from jwt import PyJWKClient
+from typing import Any, Dict, List
+from firebase_admin import _utils
+
+_APP_CHECK_ATTRIBUTE = '_app_check'
+
+def _get_app_check_service(app) -> Any:
+    return _utils.get_app_service(app, _APP_CHECK_ATTRIBUTE, _AppCheckService)
+
+# should i accept an app (design doc doesn't have one) or just always make it none
+def verify_token(token: str, app=None) -> Dict[str, Any]:
+    return _get_app_check_service(app).verify_token(token)
+
+class _AppCheckService:
+    """Service class that implements Firebase App Check functionality."""
+    
+    _APP_CHECK_GCP_API_URL = "https://firebaseappcheck.googleapis.com"
+    _APP_CHECK_BETA_JWKS_RESOURCE = "/v1beta/jwks" 
+
+
+    def __init__(self, app):
+        # the verification method should go in the service
+        project_id = app.project_id
+        if not project_id:
+            raise ValueError(
+                'Project ID is required to access App Check service. Either set the '
+                'projectId option, or use service account credentials. Alternatively, set the '
+                'GOOGLE_CLOUD_PROJECT environment variable.')
+        # Unsure what I should include in this constructor, or even if I should include one
+    
+    @classmethod
+    def verify_token(self, token: str) -> Dict[str, Any]:
+        if token is None:
```

Comment:
Do I even need this check or would it fail on compile if a str wasn't passed in?

Reply:
We need to check the token here and throw if it is not a valid string (the type hints won't validate it). You can check how we validate the string in Node.js as an example https://github.com/firebase/firebase-admin-node/blob/318f0e43d65f47dfae81decf5166ddd198d99680/src/app-check/token-verifier.ts#L53

Reply:
Added a `_Validators` class